README.md
FileSaver.js
============

FileSaver.js implements the HTML5 W3C `saveAs()` FileSaver interface in browsers that do
not natively support it. There is a [FileSaver.js demo][1] that demonstrates saving
various media types.

FileSaver.js is the solution to saving files on the client-side, and is perfect for
webapps that need to generate files, or for saving sensitive information that shouldn't be
sent to an external server.

Looking for `canvas.toBlob()` for saving canvases? Check out
[canvas-toBlob.js][2] for a cross-browser implementation.

Supported browsers
------------------

| Browser        | Constructs as | Filenames    | Max Blob Size | Dependencies |
| -------------- | ------------- | ------------ | ------------- | ------------ |
| Firefox 20+    | Blob          | Yes          | 800 MiB       | None         |
| Firefox < 20   | data: URI     | No           | n/a           | [Blob.js](https://github.com/eligrey/Blob.js) |
| Chrome         | Blob          | Yes          | [500 MiB][3]  | None         |
| Chrome for Android | Blob      | Yes          | [500 MiB][3]  | None         |
| IE 10+         | Blob          | Yes          | 600 MiB       | None         |
| Opera 15+      | Blob          | Yes          | 500 MiB       | None         |
| Opera < 15     | data: URI     | No           | n/a           | [Blob.js](https://github.com/eligrey/Blob.js) |
| Safari 6.1+*   | Blob          | No           | ?             | None         |
| Safari < 6     | data: URI     | No           | n/a           | [Blob.js](https://github.com/eligrey/Blob.js) |

Feature detection is possible:

```js
try {
    var isFileSaverSupported = !!new Blob;
} catch (e) {}
```

### IE < 10

It is possible to save text files in IE < 10 without Flash-based polyfills.
See [ChenWenBrian and koffsyrup's `saveTextAs()`](https://github.com/koffsyrup/FileSaver.js#examples) for more details.

### Safari 6.1+

Blobs may be opened instead of saved sometimes—you may have to direct your Safari users to manually
press <kbd>⌘</kbd>+<kbd>S</kbd> to save the file after it is opened. Using the `application/octet-stream` MIME type to force downloads [can cause issues in Safari](https://github.com/eligrey/FileSaver.js/issues/12#issuecomment-47247096).

### iOS

saveAs must be run within a user interaction event such as onTouchDown or onClick; setTimeout will prevent saveAs from triggering. Due to restrictions in iOS saveAs opens in a new window instead of downloading, if you want this fixed please [tell Apple](https://bugs.webkit.org/show_bug.cgi?id=102914) how this bug is affecting you.

Syntax
------

```js
FileSaver saveAs(Blob data, DOMString filename, optional Boolean disableAutoBOM)
```

Pass `true` for `disableAutoBOM` if you don't want FileSaver.js to automatically provide Unicode text encoding hints (see: [byte order mark](https://en.wikipedia.org/wiki/Byte_order_mark)).

Examples
--------

### Saving text

```js
var blob = new Blob(["Hello, world!"], {type: "text/plain;charset=utf-8"});
saveAs(blob, "hello world.txt");
```

The standard W3C File API [`Blob`][4] interface is not available in all browsers.
[Blob.js][5] is a cross-browser `Blob` implementation that solves this.

### Saving a canvas

```js
var canvas = document.getElementById("my-canvas"), ctx = canvas.getContext("2d");
// draw to canvas...
canvas.toBlob(function(blob) {
    saveAs(blob, "pretty image.png");
});
```

Note: The standard HTML5 `canvas.toBlob()` method is not available in all browsers.
[canvas-toBlob.js][6] is a cross-browser `canvas.toBlob()` that polyfills this.


![Tracking image](https://in.getclicky.com/212712ns.gif)

  [1]: http://eligrey.com/demos/FileSaver.js/
  [2]: https://github.com/eligrey/canvas-toBlob.js
  [3]: https://code.google.com/p/chromium/issues/detail?id=375297
  [4]: https://developer.mozilla.org/en-US/docs/DOM/Blob
  [5]: https://github.com/eligrey/Blob.js
  [6]: https://github.com/eligrey/canvas-toBlob.js

Contributing
------------

The `FileSaver.js` distribution file is compiled with Uglify.js like so:

```bash
npm run build
```

Please make sure you build a production version before submitting a pull request.

Bower Installation
------------------

Please see the [this repo](http://github.com/Teleborder/FileSaver.js) for a bower-compatible fork of FileSaver.js, available under the package name `file-saver.js`.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 0.2.0 - 2016-02-25
First version, forked from [eligray/FileSaver.js] to enable NPM
distribution and possible future improvements.

## Added
- package.json for (private) npm distribution
- NPM script for consistent builds
- This changelog

[eligray/FileSaver.js]: https://github.com/eligrey/FileSaver.js


gulpfile.js
/*jslint node: true */
(function () {
    'use strict';

    var gulp = require('gulp'),
        del = require('del'),
        plugins = require('gulp-load-plugins')();

    gulp.task('clean', function (cb) {
        del(['dist'], cb);
    });

    gulp.task('lint', function () {
        return gulp.src('src/**/*.js')
            .pipe(plugins.jslint());
    });

    gulp.task('default', ['clean'], function () {
        /*
         * Load in the JS file
         */
        return gulp.src('src/**/*.js')
            /*
             * Uglify (minify) it
             */
            .pipe(plugins.uglify())
            /*
             * Annotate that it was minified by adding .min to filename
             */
            .pipe(plugins.rename({
                suffix: '.min'
            }))
            /*
             * Output to our distribution location
             */
            .pipe(gulp.dest('dist'));
    });
}());

Gruntfile.js
'use strict';

module.exports = gruntConfig;


function gruntConfig(grunt) {

  // Load grunt tasks automatically
  require('load-grunt-tasks')(grunt);

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);


  grunt.initConfig({

    watch: {
      sources: {
        files: ['src/*'],
        tasks: ['concat:sources', 'uglify:dist']
      }
    },


    // Make sure code styles are up to par and there are no obvious mistakes
    jshint: {
      options: {
        jshintrc: '.jshintrc'
        //reporter: require('jshint-stylish')
      },
      all: {
        src: [
          'Gruntfile.js',
          'src/*.js'
        ]
      },
      test: {
        options: {
          jshintrc: 'test/.jshintrc'
        },
        src: ['test/spec/{,*/}*.js']
      }
    },


    concat: {
      options: {
        separator: '\n'
      },
      sources: {
        src: [
          'src/header.txt',
          'src/handle-ie.js',
          'src/save-as.js',
          'src/file-saver.js',
          'src/utils.js',
          'src/footer.txt'
        ],
        dest: 'FileSaver.js'
      }
    },


    uglify: {
      options: {
        banner: '',
        footer: '',
        beautify: false,
        mangle: false,
        compress: false,
        quoteStyle: 1
      },
      dist: {
        files: {
          'FileSaver.min.js' : [ 'FileSaver.js' ]
        }
      }
    }

  });

  grunt.registerTask('default', ['build']);

  grunt.registerTask('build', ['concat', 'uglify']);

}


src/file-saver.js
var FSProto = FileSaver.prototype;
FSProto.abort = fileSaverAbort;
FSProto.INIT = 0;
FSProto.WRITING = 1;
FSProto.DONE = 2;
FSProto.readyState = FSProto.INIT;
FSProto.error = null;
FSProto.onwritestart = null;
FSProto.onprogress = null;
FSProto.onwrite = null;
FSProto.onabort = null;
FSProto.onerror = null;
FSProto.onwriteend = null;


function FileSaver(blob, name) {
  blob = autoBom(blob);

  // First try a.download, then web filesystem, then object URLs
  var filesaver = this;
  var type = blob.type;
  var blobChanged = false;
  var objectUrl;
  var targetView;
  var createIfNotFound = {create: true, exclusive: false};

  filesaver.readyState = filesaver.INIT;
  name = name || 'download';

  if ( canUseSaveLink ) {
    return saveUsingLinkElement();
  }

  fixChromeSaveableType();
  fixWebKitDownload();


  if ( ! reqFs) {
    return saveUsingObjectURLs();
  }

  return saveUsingFyleSystem();


  ////////////////

  function saveUsingLinkElement() {
    objectUrl = getURL().createObjectURL(blob);
    saveLinkElement.href = objectUrl;
    saveLinkElement.download = name;

    setImmediate(function () {
      triggerClickOnSaveLink();
      dispatchAll(filesaver);
      revoke(objectUrl);
      filesaver.readyState = filesaver.DONE;
    }, 0);
  }

  function fixChromeSaveableType() {
    /*
     * Object and web filesystem URLs have a problem saving in Google Chrome when
     * viewed in a tab, so I force save with application/octet-stream
     * http://code.google.com/p/chromium/issues/detail?id=91158
     * Update: Google errantly closed 91158, I submitted it again:
     * https://code.google.com/p/chromium/issues/detail?id=389642
     */
    if (view.chrome && type && type !== forceSaveableType) {
      var slice = blob.slice || blob.webkitSlice;
      blob = slice.call(blob, 0, blob.size, forceSaveableType);
      blobChanged = true;
    }
  }

  function fixWebKitDownload() {
    // Since I can't be sure that the guessed media type will trigger a download
    // in WebKit, I append .download to the filename.
    // https://bugs.webkit.org/show_bug.cgi?id=65440
    if (webkitReqFs && name !== 'download') {
      name += '.download';
    }

    if (type === forceSaveableType || webkitReqFs) {
      targetView = view;
    }
  }


  function saveUsingFyleSystem() {
    fsMinSize += blob.size;

    reqFs(view.TEMPORARY, fsMinSize, abortable(getFyleSystem), saveUsingObjectURLs);

    ////////////

    function getFyleSystem(fs) {
      fs.root.getDirectory('temp', createIfNotFound, abortable(getTempDirectory), saveUsingObjectURLs);
    }

    function getTempDirectory(dir) {
      dir.getFile(name, { create: false }, abortable(getExistentFileForRemove), abortable(existentFileNotFound) );

      /////////////

      function getExistentFileForRemove(file) {
        // delete file if it already exists
        file.remove(function() {
          save();
        });
      }

      function existentFileNotFound(ex) {
        if (ex.name === 'NotFoundError') {
          save();
        } else {
          saveUsingObjectURLs();
        }
      }


      function save() {
        dir.getFile(name, createIfNotFound, abortable(getFileForWrite), saveUsingObjectURLs);
      }

      function getFileForWrite(file) {
        file.createWriter(abortable(createWriter), saveUsingObjectURLs);

        ////////

        function createWriter(writer) {
          writer.onwriteend = onWriterEnd;
          writer.onerror = onError;

          bindEventsToWriter();

          writer.write(blob);
          filesaver.abort = onAbort;
          filesaver.readyState = filesaver.WRITING;

          ////////////

          function onWriterEnd(event) {
            execSave(file.toURL(), file, event);
          }

          function onError() {
            var error = writer.error;
            if (error.code !== error.ABORT_ERR) {
              saveUsingObjectURLs();
            }
          }

          function bindEventsToWriter() {
            'writestart progress write abort'.split(' ').forEach(function (event) {
              writer['on' + event] = filesaver['on' + event];
            });
          }

          function onAbort() {
            writer.abort();
            filesaver.readyState = filesaver.DONE;
          }
        }
      }
    }
  }

  // on any filesys errors revert to saving with object URLs
  function saveUsingObjectURLs() {
    // don't create more object URLs than needed
    if (blobChanged || !objectUrl) {
      objectUrl = getURL().createObjectURL(blob);
    }

    execSave(objectUrl);
  }


  function execSave(objectUrl, file, event) {
    var newTab = view.open(objectUrl, '_blank');
    if ( newTab === undefined ) {
      /*
       * Apple do not allow window.open
       * see http://bit.ly/1kZffRI
       */
      view.location.href = objectUrl;
    }


    filesaver.readyState = filesaver.DONE;
    if( ! event ) {
      dispatchAll(filesaver);
    } else {
      dispatch(filesaver, 'writeend', event);
    }

    revoke(file || objectUrl);
  }


  function abortable(func) {
    return function () {
      if (filesaver.readyState !== filesaver.DONE) {
        return func.apply(this, arguments);
      }
    };
  }

}

function fileSaverAbort() {
  var filesaver = this;
  filesaver.readyState = filesaver.DONE;
  dispatch(filesaver, 'abort');
}


// TODO apply dispatch to FileSaver prototype
function dispatchAll(filesaver) {
  dispatch(filesaver, 'writestart progress write writeend'.split(' '));
}


function dispatch(filesaver, eventTypes, event) {
  eventTypes = [].concat(eventTypes);
  var i = eventTypes.length;
  while (i--) {
    var listener = filesaver['on' + eventTypes[i]];
    if (typeof listener === 'function') {
      try {
        listener.call(filesaver, event || filesaver);
      } catch (ex) {
        throwOutside(ex);
      }
    }
  }
}


FileSaver.js
/* FileSaver.js
 * A saveAs() FileSaver implementation.
 * 1.1.20151003
 *
 * By Eli Grey, http://eligrey.com
 * License: MIT
 *   See https://github.com/eligrey/FileSaver.js/blob/master/LICENSE.md
 */

/*global self */
/*jslint bitwise: true, indent: 4, laxbreak: true, laxcomma: true, smarttabs: true, plusplus: true */

/*! @source http://purl.eligrey.com/github/FileSaver.js/blob/master/FileSaver.js */

var saveAs = saveAs || (function(view) {
	"use strict";
	// IE <10 is explicitly unsupported
	if (typeof navigator !== "undefined" && /MSIE [1-9]\./.test(navigator.userAgent)) {
		return;
	}
	var
		  doc = view.document
		  // only get URL when necessary in case Blob.js hasn't overridden it yet
		, get_URL = function() {
			return view.URL || view.webkitURL || view;
		}
		, save_link = doc.createElementNS("http://www.w3.org/1999/xhtml", "a")
		, can_use_save_link = "download" in save_link
		, click = function(node) {
			var event = new MouseEvent("click");
			node.dispatchEvent(event);
		}
		, is_safari = /Version\/[\d\.]+.*Safari/.test(navigator.userAgent)
		, webkit_req_fs = view.webkitRequestFileSystem
		, req_fs = view.requestFileSystem || webkit_req_fs || view.mozRequestFileSystem
		, throw_outside = function(ex) {
			(view.setImmediate || view.setTimeout)(function() {
				throw ex;
			}, 0);
		}
		, force_saveable_type = "application/octet-stream"
		, fs_min_size = 0
		// See https://code.google.com/p/chromium/issues/detail?id=375297#c7 and
		// https://github.com/eligrey/FileSaver.js/commit/485930a#commitcomment-8768047
		// for the reasoning behind the timeout and revocation flow
		, arbitrary_revoke_timeout = 500 // in ms
		, revoke = function(file) {
			var revoker = function() {
				if (typeof file === "string") { // file is an object URL
					get_URL().revokeObjectURL(file);
				} else { // file is a File
					file.remove();
				}
			};
			if (view.chrome) {
				revoker();
			} else {
				setTimeout(revoker, arbitrary_revoke_timeout);
			}
		}
		, dispatch = function(filesaver, event_types, event) {
			event_types = [].concat(event_types);
			var i = event_types.length;
			while (i--) {
				var listener = filesaver["on" + event_types[i]];
				if (typeof listener === "function") {
					try {
						listener.call(filesaver, event || filesaver);
					} catch (ex) {
						throw_outside(ex);
					}
				}
			}
		}
		, auto_bom = function(blob) {
			// prepend BOM for UTF-8 XML and text/* types (including HTML)
			if (/^\s*(?:text\/\S*|application\/xml|\S*\/\S*\+xml)\s*;.*charset\s*=\s*utf-8/i.test(blob.type)) {
				return new Blob(["\ufeff", blob], {type: blob.type});
			}
			return blob;
		}
		, FileSaver = function(blob, name, no_auto_bom, failureCallback) {
			if (!no_auto_bom) {
				blob = auto_bom(blob);
			}
			// First try a.download, then web filesystem, then object URLs
			var
				  filesaver = this
				, type = blob.type
				, blob_changed = false
				, object_url
				, target_view
				, dispatch_all = function() {
					dispatch(filesaver, "writestart progress write writeend".split(" "));
				}
				// on any filesys errors revert to saving with object URLs
				, fs_error = function() {
					if (target_view && is_safari && typeof FileReader !== "undefined") {
						// Safari doesn't allow downloading of blob urls
						var reader = new FileReader();
						reader.onloadend = function() {
							var base64Data = reader.result;
							target_view.location.href = "data:attachment/file" + base64Data.slice(base64Data.search(/[,;]/));
							filesaver.readyState = filesaver.DONE;
							dispatch_all();
						};
						reader.readAsDataURL(blob);
						filesaver.readyState = filesaver.INIT;
						return;
					}
					// don't create more object URLs than needed
					if (blob_changed || !object_url) {
						object_url = get_URL().createObjectURL(blob);
					}
					// if url callback specified then assume caller will handle download rather than changing the current window location.
					if (failureCallback) {
						failureCallback(object_url);
					}
					else {
						if (target_view) {
							target_view.location.href = object_url;
						} else {
							var new_tab = view.open(object_url, "_blank");
							if (new_tab == undefined && is_safari) {
								//Apple do not allow window.open, see http://bit.ly/1kZffRI
								view.location.href = object_url
							}
						}
					}
					filesaver.readyState = filesaver.DONE;
					dispatch_all();
					if (!failureCallback) { // don't revoke if the object_url is passed into callback.
						revoke(object_url);
					}
				}
				, abortable = function(func) {
					return function() {
						if (filesaver.readyState !== filesaver.DONE) {
							return func.apply(this, arguments);
						}
					};
				}
				, create_if_not_found = {create: true, exclusive: false}
				, slice
			;
			filesaver.readyState = filesaver.INIT;
			if (!name) {
				name = "download";
			}
			if (can_use_save_link) {
				object_url = get_URL().createObjectURL(blob);
				setTimeout(function() {
					save_link.href = object_url;
					save_link.download = name;
					click(save_link);
					dispatch_all();
					revoke(object_url);
					filesaver.readyState = filesaver.DONE;
				});
				return;
			}
			// Object and web filesystem URLs have a problem saving in Google Chrome when
			// viewed in a tab, so I force save with application/octet-stream
			// http://code.google.com/p/chromium/issues/detail?id=91158
			// Update: Google errantly closed 91158, I submitted it again:
			// https://code.google.com/p/chromium/issues/detail?id=389642
			if (view.chrome && type && type !== force_saveable_type) {
				slice = blob.slice || blob.webkitSlice;
				blob = slice.call(blob, 0, blob.size, force_saveable_type);
				blob_changed = true;
			}
			// Since I can't be sure that the guessed media type will trigger a download
			// in WebKit, I append .download to the filename.
			// https://bugs.webkit.org/show_bug.cgi?id=65440
			if (webkit_req_fs && name !== "download") {
				name += ".download";
			}
			if (type === force_saveable_type || webkit_req_fs) {
				target_view = view;
			}
			if (!req_fs) {
				fs_error();
				return;
			}
			fs_min_size += blob.size;
			req_fs(view.TEMPORARY, fs_min_size, abortable(function(fs) {
				fs.root.getDirectory("saved", create_if_not_found, abortable(function(dir) {
					var save = function() {
						dir.getFile(name, create_if_not_found, abortable(function(file) {
							file.createWriter(abortable(function(writer) {
								writer.onwriteend = function(event) {
									target_view.location.href = file.toURL();
									filesaver.readyState = filesaver.DONE;
									dispatch(filesaver, "writeend", event);
									revoke(file);
								};
								writer.onerror = function() {
									var error = writer.error;
									if (error.code !== error.ABORT_ERR) {
										fs_error();
									}
								};
								"writestart progress write abort".split(" ").forEach(function(event) {
									writer["on" + event] = filesaver["on" + event];
								});
								writer.write(blob);
								filesaver.abort = function() {
									writer.abort();
									filesaver.readyState = filesaver.DONE;
								};
								filesaver.readyState = filesaver.WRITING;
							}), fs_error);
						}), fs_error);
					};
					dir.getFile(name, {create: false}, abortable(function(file) {
						// delete file if it already exists
						file.remove();
						save();
					}), abortable(function(ex) {
						if (ex.code === ex.NOT_FOUND_ERR) {
							save();
						} else {
							fs_error();
						}
					}));
				}), fs_error);
			}), fs_error);
		}
		, FS_proto = FileSaver.prototype
		, saveAs = function(blob, name, no_auto_bom, failureCallback) {
			return new FileSaver(blob, name, no_auto_bom, failureCallback);
		}
	;
	// IE 10+ (native saveAs)
	if (typeof navigator !== "undefined" && navigator.msSaveOrOpenBlob) {
		return function(blob, name, no_auto_bom) {
			if (!no_auto_bom) {
				blob = auto_bom(blob);
			}
			return navigator.msSaveOrOpenBlob(blob, name || "download");
		};
	}

	FS_proto.abort = function() {
		var filesaver = this;
		filesaver.readyState = filesaver.DONE;
		dispatch(filesaver, "abort");
	};
	FS_proto.readyState = FS_proto.INIT = 0;
	FS_proto.WRITING = 1;
	FS_proto.DONE = 2;

	FS_proto.error =
	FS_proto.onwritestart =
	FS_proto.onprogress =
	FS_proto.onwrite =
	FS_proto.onabort =
	FS_proto.onerror =
	FS_proto.onwriteend =
		null;

	return saveAs;
}(
	   typeof self !== "undefined" && self
	|| typeof window !== "undefined" && window
	|| this.content
));
// `self` is undefined in Firefox for Android content script context
// while `this` is nsIContentFrameMessageManager
// with an attribute `content` that corresponds to the window

if (typeof module !== "undefined" && module.exports) {
  module.exports.saveAs = saveAs;
} else if ((typeof define !== "undefined" && define !== null) && (define.amd != null)) {
  define([], function() {
    return saveAs;
  });
}
